### ISSUE TITLE
Implement reading and parsing of MIDI messages

### ISSUE THREAD

**Issue by kevinjwalters** [adafruit/Adafruit_CircuitPython_MIDI#3]
Is there existing work or a plan to add support for reading and parsing of _incoming_ MIDI message?

I might want to do this in the near future. If no one else is working on it then I could provide a partial implementation of this? Is there any existing code demonstrating a good design pattern for this?

Current implementation:
- Can read MIDI messages from one or more specified channels using `receive()` method which is non-blocking for USB MIDI
- Each message is a class similar to [adafruit_bluefruit_connect](https://github.com/adafruit/Adafruit_CircuitPython_BluefruitConnect)
- Notes can also be sent with the same messages using new `send()` method
- `NoteOn`, `NoteOff`, `PolyphonicKeyPressure` (aftertouch) can be constructed with notes in text form, **C4** is middle C, **A4** is traditional A above C4 (typicalyl 440Hz ref).
- Decent number of unit tests to ensure parser works.

**Comment by ladyada** [comment]
i was going to do it but ran out of time, @tannewt suggest do it with it returning objects, not lists of bytes. sorta like how we do it here https://github.com/adafruit/Adafruit_CircuitPython_BluefruitConnect

**Comment by kevinjwalters** [comment]
Got something working using objects per message type. Will be interesting to see how that scales wrt memory usage but it's ok so far.

`
>>> while True:
...     msg = midi.read_in_port()
...     if msg is not None:
...         print(msg)
...
...
...
<NoteOn object at 200035f0>
<NoteOff object at 200036b0>
<NoteOn object at 20003e10>
<NoteOff object at 20003ed0>
<NoteOn object at 20004590>
<NoteOff object at 20004650>
<NoteOn object at 20004d20>
<NoteOff object at 20004de0>
<NoteOn object at 20001780>
<NoteOff object at 20001960>
<NoteOn object at 200025c0>
<NoteOff object at 20002660>
<NoteOn object at 20003000>
<NoteOff object at 200030c0>
<NoteOn object at 200038a0>
<NoteOff object at 20003960>
<NoteOn object at 200040a0>
<NoteOff object at 20004160>
<NoteOn object at 20004850>
<NoteOff object at 20004910>
<NoteOn object at 20004f80>
<NoteOff object at 20005040>
<NoteOn object at 20001e30>
<NoteOff object at 20001f10>
<NoteOn object at 200020f0>
<NoteOn object at 20002ac0>
<NoteOff object at 20002b80>
<NoteOn object at 20003340>
<ControlChange object at 20003400>
<ControlChange object at 200034c0>
<NoteOn object at 20003580>
<ControlChange object at 20003640>
<ControlChange object at 20003700>
`

**Comment by ladyada** [comment]
nice!

**Comment by kevinjwalters** [comment]
I've got a basic digital square wave monosynth working with:
- new midi input code
- CPX
- a breadboard with three transistors, eight resistors, three capacitors and a 3.5mm socket
- https://github.com/kevinjwalters/circuitpython-examples/blob/master/cpx/cpx-basic-square-monosynth.py

**Comment by kevinjwalters** [comment]
@jedgarpark in Discord:

> @kevinjwalters excellent. We just have CC, PitchBend, NoteOn, and NoteOff at the moment, right? Patch Change would be cool. I don't know if I'll make any controllers with Aftertouch or Channel Pressure, but those could be cool if someone did.

**Comment by kevinjwalters** [comment]
Have a look at CM messages in particular 123 as this may be what midi controllers with a panic button send. Check what Axiom sends.

**Comment by tannewt** [comment]
This is very cool! One idea we had to save memory is to only parse message types that have been imported. (An import will trigger a registration like the BLE code.) That way others can be ignored.

**Comment by kevinjwalters** [comment]
Try using travis to run the unit tests. @tannewt showed me https://github.com/adafruit/Adafruit_CircuitPython_Motor/blob/master/.travis.yml as an example of this. It uses `py.test`.

**Comment by kevinjwalters** [comment]
Almost finished just a few more tests to write, review TODO notes, and minor tidying and polishing of docs.

**Comment by kevinjwalters** [comment]
I was unaware of [NeoTrellis M4 MIDI File Synthesizer](https://learn.adafruit.com/neotrellism4-midi-file-player-synthesizer). This appears to read (and therefore parse) MIDI files so there's some strong overlap here.

**Comment by kevinjwalters** [comment]
I think I've found an issue with `pylint` not checking CircuitPython/microPython `gc` library properly. I'm using `gc.mem_free()` in an example designed purely to profile memory usage, `pylint` chokes on it with:

```
E:  8,21: Module 'gc' has no 'mem_free' member (no-member)
```

Full output in: https://travis-ci.com/adafruit/Adafruit_CircuitPython_MIDI/builds/107358306

I've suppressed it for now in that file so this is not blocking build.

**Comment by kevinjwalters** [comment]
This is the free memory as all the messages are loaded in by [midi_memorycheck.py](https://github.com/kevinjwalters/Adafruit_CircuitPython_MIDI/blob/621fded7723075dd5b1d4e1cc766b92d9b2dfaed/examples/midi_memorycheck.py) for library version https://github.com/kevinjwalters/Adafruit_CircuitPython_MIDI/tree/e3dfd0ee0c968e11d34b5c22f18b7ce0ff4d402a:

```
Adafruit CircuitPython 4.0.0-beta.5 on 2019-03-17; Adafruit CircuitPlayground Express with samd21g18
>>>
>>>
soft reboot

Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
20128
20128
13216
12464
11520
10656
9664
8880
7904
7152
6672
6224
5120
4672
4528
```

The output can vary very slightly between two values on different runs in common with other CircuitPython code, mentioned on: https://forums.adafruit.com/viewtopic.php?f=60&t=148334

**Comment by ladyada** [comment]
@tannewt any ideas on how to optimize so it can work the way you want on M0?

**Comment by tannewt** [comment]
I don't have any concrete suggestions. It would take a deeper dive into a sample program to have an idea where to look.

I'd expect this to work ok on an M0 as long as not all messages are imported. Is there an example where it isn't working?

**Comment by kevinjwalters** [comment]
> I don't have any concrete suggestions. It would take a deeper dive into a sample program to have an idea where to look.
> 
> I'd expect this to work ok on an M0 as long as not all messages are imported. Is there an example where it isn't working?

It's not really binary in the sense of works doesn't work it's more of the balance between library size and application size. I had to drop my use of `ChannelPressure` in https://github.com/kevinjwalters/circuitpython-examples/blob/master/cpx/cpx-threeosc-synth.py with recent changes. I'd already been sacrificing samples for my waveform output (note reduction of `basesamplerate`).

**Comment by tannewt** [comment]
Ya, I think that is to be expected on the M0. I don't think we should wait on any changes to the library but I'm sure we could optimize things after looking at the heap diagnostics.

**Comment by kevinjwalters** [comment]
Whilst checking out recent library bundle I noticed that` __init__.py` files are not compiled to `.mpy` files. It was apparent because of a MemoryError from a previously working application that was using pre-bundle hand `mpy-cross` compiled files.

The addition of per MIDI message (event) classes implied indidvidual files which in turn forced the conversion of `adafruit_midi.py` to `adafruit_midi/__init__.py`.

In the land of discord, @sommersoft has noted:

> i didn't even think about that during the review of your PR. the release build script actually skips all __init__.py files: https://github.com/adafruit/circuitpython-build-tools/blob/master/circuitpython_build_tools/build.py#L120 -> https://github.com/adafruit/circuitpython-build-tools/blob/master/circuitpython_build_tools/build.py#L33 could be worth starting the conversation to change that, if you don't want to move it elsewhere...

And @deshipu:

> you shouldn't be putting any substantial code in __init__.py anywyas

The line referred to is an intentional (uncommented) exclusion of:

```
IGNORE_PY = ["setup.py", "conf.py", "__init__.py"]
```

It would be useful to know the intention/purpose of that from [build.py](https://github.com/adafruit/circuitpython-build-tools/blob/master/circuitpython_build_tools/build.py#L33) ? If it's to avoid compiling non library files that it seems wrong to exclude `__init__.py`? It would also be better if it targetted pathed files as I suspect it's trying to avoid pointless compilation of `./setup.py` and `./docs/conf.py`? @tannewt any thoughts on this?

I also see some special casing for `__init__.py` at https://github.com/adafruit/circuitpython-build-tools/blob/master/circuitpython_build_tools/build.py#L168 but I still haven't figured out exactly how the directories are being traversed particularly because there could be some directory depth on more complex libraries.

**Comment by kevinjwalters** [comment]
Quick survey of non-zero length `__init__.py` files from `adafruit-circuitpython-bundle-4.x-mpy-20190425` bundle:

```
adafruit_midi/__init__.py                           9186
adafruit_imageload/__init__.py                      2249
adafruit_bluefruit_connect/__in...                  1930
adafruit_ble/__init__.py                            1819
adafruit_hid/__init__.py                            1380
adafruit_thermal_printer/__init...                   982
adafruit_tinylora/__init__.py                        177
adafruit_rgb_display/__init__.py                     106
adafruit_apds9960/__init__.py                         43
```

**Comment by kevinjwalters** [comment]
Will put in a separate ticket if any work needs to be done on __init__.py size tweaks or compilation of __init__.py (probably goes in https://github.com/adafruit/circuitpython-build-tools). #12 work also a separate ticket.

**Comment by kevinjwalters** [comment]
Last issue mentioned in comments now ticketed as https://github.com/adafruit/circuitpython-build-tools/issues/29
